cli/lib/cfg/index.js

```diff
@@ -1,64 +1,147 @@
-        return {};
-      }
-    })
-    .then(function (d) {
-      d.path=p;
-      set(d);
-      return curCfg;
-    })
+
+  try {
+    set(require(configLocation));
+  } catch (e) {
+    log.warn('Config file not found. Will initialise new config at %s', p);
+    set({});
+  }
+
+  // Store the config path in the config - TODO is this required?
+  curCfg.path = configLocation;
+
+  return Promise.resolve(curCfg);
 }
-function setKey(key, val) {
+
+/**
+ * Set a given key to a given value in the config.
+ * The key can be a nested string e.g 'paths.cordova'
+ * @param {String} key
+ * @param {String} val
+ */
+exports.setKey = function (key, val) {
+  debug('setting key %s to %j', key, val);
   _.set(curCfg, key, val);
-}
+};
 
-function getKey(key, def) {
-  if (!curCfg) {
-    throw ("Config is not loaded and getKey is called.");
-  }
-  return _.get(curCfg, key, def);
+/**
+ * Retrieve a key from the config. Can be nested e.g 'paths.cordova'
+ * @param {String} key
+ * @param {String} defaultValue
+ */
+function getKey(key, defaultValue) {
+  assert(curCfg, 'Config file not loaded. Please ensure you called config.load()');
+
+  return _.get(curCfg, key, defaultValue);
 }
-function set(data) {
+
+/**
+ * Replace the in memory config with the one passed
+ * @param {Object} data
+ */
+const set = exports.set = function (data) {
+  assert(_.isObject(data), 'Config must be an Object');
+  debug('replacing config object with %j', data);
   curCfg = data;
 }
-function get() {
+
+/**
+ * Return the loaded config
+ * @return {Object}
+ */
+const get = exports.get = function () {
   return curCfg;
 }
 
-function save(path) {
-  var p = getPath(path);
-  return P.promisify(fs.writeFile)(p, JSON.stringify(curCfg, null, 2));
+/**
+ * Save the config at the specified path. Defaults to current working directory.
+ * @param {String} [path]
+ */
+const save = exports.save = function (path) {
+  assert(curCfg, 'Config file not loaded. Please ensure you called config.load()');
+
+  const resolvedPath = getPath(path);
+
+  debug('saving the current config to disk at path %s with content %j', resolvedPath, curCfg)
+
+  return new Promise((resolve) => {
+    writeFileSync(resolvedPath, JSON.stringify(curCfg, null, 2))
+    resolve();
+  });
 }
-function getPath(path) {
-  if (!path) {
-    path = process.cwd();
-  }
-  return require("path").join(path, cfgFile);
-}
\ No newline at end of file
+
+/**
+ * Returns the location of the Cordova executable for the given project
+ * @return {String}
+ */
+exports.getCordovaLocation = function () {
+  return getKey('cordova.location', 'cordova');
+};
+
+
+/**
+ * Returns the server URL where hcp server is located
+ * @return {String}
+ */
+exports.getServerUrl = function () {
+  return getKey('serverUrl');
+};
+
+/**
+ * Creates a new config for a project, saves it to disk, then returns it.
+ * @param   {Object} args
+ * @return  {Promise<Object>}
+ */
+exports.generateConfig = function (args) {
```

Comment:
This is the main new addition here. It will generate a new config. Should probably add some `assert` calls?